Add paragraph about `number_of_` to the styleguide. (#3305)
oscar-system/Oscar.jl: docs/src/DeveloperDocumentation/styleguide.md

  also being consistent.
  For compatibility with standard Julia, while staying consistent internally,
  we also provide aliases (using `AbstractAlgebra.@alias`) for various standard
  Julia functions, e.g. `is_one` as alias for `isone`.
- A function returning the number of some things should be named `number_of_things`.
  For some very common things, like the number of generators, we additionally

>   provide an alias `ngens` for `number_of_generators`. These aliases should be

Reviewer: ```suggestion
  provide a shorter alias, e.g. `ngens` for `number_of_generators`. These aliases should be
```

  short (< 15 chararacters) and without underscores, and one should only use

>   them inside of functions, not in the user interface and documentation.

Reviewer: I think this is not correct anymore, but I am not sure what the correct new description would be. Perhaps @lkastner @benlorenz @micjoswig could propose something?

In particular I am not quite sure myself. So the short aliases now start with `n_`, but we keep `number_of_`? Or do we plan to phase out `number_of_` in favor of `n_` (I'd be fine with that, too, just need to know what the plan is).

Thanks!

>   them inside of functions, not in the user interface and documentation.

Reviewer 2: ```suggestion
- A function returning the number of some things should be named `number_of_things`,
  alternatively it can be named `n_things` with an alias to `number_of_things`.
  The preferred style should be consistent throughout the corresponding part.
  For some very common things, like the number of generators, we additionally
  provide a shorter alias, e.g. `ngens` for `number_of_generators`. These aliases should be
  very short and without underscores.
```
We keep `number_of_` and allow the use of `n_` with a `number_of_` alias as well.

The short and abbreviated aliases `ngens`, `nvars`, `ncols` and `nrows` also stay as they are. But I do not think we can expect developers to use them only inside functions but not for user-facing code so I removed that part.

  also being consistent.
  For compatibility with standard Julia, while staying consistent internally,
  we also provide aliases (using `AbstractAlgebra.@alias`) for various standard
  Julia functions, e.g. `is_one` as alias for `isone`.
- A function returning the number of some things should be named `number_of_things`,

>   alternatively it can be named `n_things` with an alias to `number_of_things`.

Author: This reads to me as if everything should have a `number_of_things` function and may additionally have a `n_things` alias. Is this what you intended to say here? I understood it before that every `number_of_things` function should have an `n_things` alias.

Reviewer 2: Everything should be named either `number_of_things` or `n_things`. If it is named `n_things` (and documented like that) then it should have an alias `number_of_things`. If it is named `number_of_` it may have an alias `n_`.

I did not add `n_` aliases for all `number_of` functions.

Author: Ah, I didn't notice you swapped the alias direction of only some functions in #3364. I would find it way more consistent to have all aliases the same there, i.e. `number_of_` is the default, and then `n_` is the alias. But that's a different story. The formulation here now matches the reality of the implementation

Reviewer: I think this paragraph is fine now -- at least it seems to match what @benlorenz described and I am fine with it, too.

  The preferred style should be consistent throughout the corresponding part.
  For some very common things, like the number of generators, we additionally
  provide a shorter alias, e.g. `ngens` for `number_of_generators`. These aliases should be

>   very short and without underscores.

Author: ```suggestion
  very short and without underscores, and one should only use
  them inside of functions, not in the user interface and documentation.
```
I think this is important to state: For consistency, I think we should attach docstrings not to `ngens` but to one of the other two, and the docstring should contain the signature of the "long form". It may additionally state the existence of the `ngens` alias, but that should not be too dominantly to not confuse users with three names for a single function.

Reviewer 2: I don't really like the "developers may use ngens internally but users should not be exposed to it too much" approach. Julia makes it very easy to look at the code with `@less` or similar functions.

There are about 20 `ngens` calls in the book as well.

Reviewer 3: Since the changes are on a branch of a private repo (which I find unnecessarily tedious), I just write comments.

* `number_of` or `n_` but only one version per section. Example: `n_` is used throughout the polyhedral geometry section.
* `number_of` should be available as alias if the function is written as `n_`
* `n_` may be used as an alias if the function is written as `number_of`
* `ngens` should become `n_gens` then

I agree with @benlorenz that it does not make much sense to distinguish between "internal" and "external".  After all, this is an open source project.

Reviewer: So it seems we agree at least in so far as: there should always be a `number_of_BLA` and there *may* be also a `n_BLA` (ignoring the technical details about e.g. which one may or may not be an "alias" (as defined via `@alias`) of which one for the moment).

But note that renaming `ngens` to `n_gens` would not quite fit: since we have `number_of_generators` it would have to be called `n_generators`... Moreover we use `ngens` and `gens` in ~1000 spots in Oscar.jl alone, and as Benjamin pointed out, it is used 20 times in book.

This is why in January a dozen of us agreed after multiple discussions over two weeks to allow certain "ultra short" aliases being grandfathered in... besides `ngens` that's also `nrows` and `ncols` and probably more (e.g. perhaps `nv` or `ne` for Graphs.jl compatibility?! (I am not a fan of those names, but if someone thinks this is useful to have, I don't want to block them?).

That said, I have no personal attachment to `ngens` and could easily adapt to using `n_gens` or even `n_generators`, but my point here is that this is a big change requiring a lot of work (with changes over multiple packages, once again carefully coordinated); the `n_gens` variant still is not really consistent; and in the end we can't get rid of `ngens` for now anyway. So is it really worth the trouble at this point?

Reviewer 3: I agree with @fingolfin and @benlorenz .

That is, we have `n_things` and `number_of_things` for most things, but for historical and workload reasons the odd `ngens` is fine, too.  No `n_gens` then.

Further, as @benlorenz pointed out: no distinction between "interior use" and "interface".

@lgoettgens : Personally, I do not need `number_of_generators`, but if it makes you happy, this could be an alias to `ngens`.
